protoc-gen-gogo/generator/generator.go

```diff
@@ -2178,6 +2194,31 @@ func (g *Generator) generateMessage(message *Descriptor) {
 				log.Printf("don't know how to generate constant for %s", fieldname)
 				continue
 			}
+
+			// hunt down the actual enum corresponding to the default
+			var enumValue *descriptor.EnumValueDescriptorProto
+			for _, ev := range enum.Value {
+				// TODO: This could probably be preindexed but the performance
+				// benefit might not outweigh keeping tracking of the uniqueness
+				// across package boundaries.
+				if def == ev.GetName() {
+					enumValue = ev
+				}
+			}
+
+			if enumValue != nil {
+				if gogoproto.IsEnumValueCustomName(enumValue) {
+					def = gogoproto.GetEnumValueCustomName(enumValue)
+				}
+			} else {
+				// This condition should be rare or impossible. Could be caused
+				// by a compiler bug that didn't ensure that default enum value
+				// is define for the type.
+				g.Fail(fmt.Sprintf("could not resolve default enum value for %v.%v",
+					g.DefaultPackageName(obj), def))
+
+			}
+
 			if gogoproto.EnabledGoEnumPrefix(enum.file, enum.EnumDescriptorProto) {
```

Comment:
I actually gave this a try but ran into some odd consistency issues. 

Basically, the complexity here is that a Go namespace and a Protobuf namespace needs to be maintained. It is not wildly complex but would have made this PR much more complex. An example of the problem is that enum value search loop at line 2200. Really, each type needs to produce its Go name and Protobuf name, qualified to the file scope. This led me down the path of an interface change, which is much too big of a change when adding a small feature.

If you don't mind, I'd prefer to leave that to a refactoring, which I might look at undertaking if I do more work in here.